File: internal/orchestration/router.go
package orchestration

import "strings"

// RoutingResult pairs an agent slug with its match score for a task.
type RoutingResult struct {
	AgentSlug string
	Score     float64
}

type agentRegistration struct {
	slug   string
	skills []SkillDeclaration
}

// TaskRouter routes tasks to agents based on skill matching.
type TaskRouter struct {
	agents map[string]*agentRegistration
}

// NewTaskRouter returns an empty TaskRouter.
func NewTaskRouter() *TaskRouter {
	return &TaskRouter{agents: make(map[string]*agentRegistration)}
}

// RegisterAgent adds or replaces an agent's skill registration.
func (r *TaskRouter) RegisterAgent(slug string, skills []SkillDeclaration) {
	r.agents[slug] = &agentRegistration{slug: slug, skills: skills}
}

// UnregisterAgent removes an agent from the router.
func (r *TaskRouter) UnregisterAgent(slug string) {
	delete(r.agents, slug)
}

// ScoreMatch returns a 0-1 score for how well agentSlug can handle task.
// For each required skill, the best matching agent skill (sim * proficiency)
// is found; scores are averaged. Skills with no match above 0.3 contribute 0.
func (r *TaskRouter) ScoreMatch(agentSlug string, task TaskDefinition) float64 {
	reg, ok := r.agents[agentSlug]
	if !ok || len(task.RequiredSkills) == 0 {
		return 0
	}

	total := 0.0
	for _, required := range task.RequiredSkills {
		best := 0.0
		for _, skill := range reg.skills {
			sim := similarity(required, skill.Name)
			score := sim * skill.Proficiency
			if score > best {
				best = score
			}
		}
		if best >= 0.3 {
			total += best
		}
	}
	return total / float64(len(task.RequiredSkills))
}

// FindBestAgent returns the agent with the highest score for the task,
// or nil if no agent scores above 0.
func (r *TaskRouter) FindBestAgent(task TaskDefinition) *RoutingResult {
	results := r.FindCapableAgents(task)
	if len(results) == 0 {
		return nil
	}
	best := results[0]
	for _, rr := range results[1:] {
		if rr.Score > best.Score {
			best = rr
		}
	}
	return &best
}

// FindCapableAgents returns all agents with a score > 0, sorted descending.
func (r *TaskRouter) FindCapableAgents(task TaskDefinition) []RoutingResult {
	var out []RoutingResult
	for slug := range r.agents {
		score := r.ScoreMatch(slug, task)
		if score > 0 {
			out = append(out, RoutingResult{AgentSlug: slug, Score: score})
		}
	}
	// Sort descending by score.
	for i := 0; i < len(out); i++ {
		for j := i + 1; j < len(out); j++ {
			if out[j].Score > out[i].Score {
				out[i], out[j] = out[j], out[i]
			}
		}
	}
	return out
}

// similarity computes the Dice coefficient bigram similarity between two strings.
// Returns a value in [0, 1].
func similarity(a, b string) float64 {
	a = strings.ToLower(a)
	b = strings.ToLower(b)

	if a == b {
		return 1.0
	}
	if len(a) < 2 || len(b) < 2 {
		// For single-char strings, use exact match.
		if a == b {
			return 1.0
		}
		return 0.0
	}

	bigramsA := bigrams(a)
	bigramsB := bigrams(b)

	intersection := 0
	// Count matching bigrams (each bigram in A matched against B).
	usedB := make([]bool, len(bigramsB))
	for _, bg := range bigramsA {
		for i, bgB := range bigramsB {
			if !usedB[i] && bg == bgB {
				intersection++
				usedB[i] = true
				break
			}
		}
	}

	return float64(2*intersection) / float64(len(bigramsA)+len(bigramsB))
}

// bigrams returns the list of two-character substrings in s.
func bigrams(s string) []string {
	runes := []rune(s)
	if len(runes) < 2 {
		return nil
	}
	result := make([]string, len(runes)-1)
	for i := 0; i < len(runes)-1; i++ {
		result[i] = string(runes[i : i+2])
	}
	return result
}

File: cli/src/orchestration/router.ts
/**
 * Expertise-based task routing.
 * Matches agent skills to task requirements using fuzzy string matching
 * weighted by proficiency scores.
 */

import type { SkillDeclaration, TaskDefinition } from './types.js';

interface AgentRegistration {
  slug: string;
  skills: SkillDeclaration[];
}

/**
 * Simple normalized string similarity (Dice coefficient on bigrams).
 * Returns 0-1, where 1 is a perfect match.
 */
function similarity(a: string, b: string): number {
  const sa = a.toLowerCase().trim();
  const sb = b.toLowerCase().trim();
  if (sa === sb) return 1;
  if (sa.length < 2 || sb.length < 2) return 0;

  const bigramsA = new Set<string>();
  for (let i = 0; i < sa.length - 1; i++) bigramsA.add(sa.slice(i, i + 2));

  const bigramsB = new Set<string>();
  for (let i = 0; i < sb.length - 1; i++) bigramsB.add(sb.slice(i, i + 2));

  let intersection = 0;
  for (const bg of bigramsA) {
    if (bigramsB.has(bg)) intersection++;
  }

  return (2 * intersection) / (bigramsA.size + bigramsB.size);
}

export class TaskRouter {
  private agents: Map<string, AgentRegistration> = new Map();

  /** Register an agent with its skill declarations. */
  registerAgent(agentSlug: string, skills: SkillDeclaration[]): void {
    this.agents.set(agentSlug, { slug: agentSlug, skills });
  }

  /** Unregister an agent. */
  unregisterAgent(agentSlug: string): void {
    this.agents.delete(agentSlug);
  }

  /**
   * Score a single agent against a task.
   * For each required skill, find the best matching agent skill (fuzzy match).
   * Score = average of best matches * agent proficiency average.
   * Agents with 0 matching skills (all similarities below threshold) get 0.
   */
  scoreMatch(agentSlug: string, task: TaskDefinition): number {
    const reg = this.agents.get(agentSlug);
    if (!reg) return 0;

    const required = task.requiredSkills;
    // If task requires no skills, any agent can handle it
    if (required.length === 0) return 1;

    const agentSkills = reg.skills;
    if (agentSkills.length === 0) return 0;

    const MATCH_THRESHOLD = 0.3;
    let matchSum = 0;
    let matchCount = 0;

    for (const reqSkill of required) {
      let bestSim = 0;
      let bestProf = 0;
      for (const agentSkill of agentSkills) {
        const sim = similarity(reqSkill, agentSkill.name);
        if (sim > bestSim) {
          bestSim = sim;
          bestProf = agentSkill.proficiency;
        }
      }
      if (bestSim >= MATCH_THRESHOLD) {
        matchSum += bestSim * bestProf;
        matchCount++;
      }
    }

    // If no skills matched above threshold, return 0
    if (matchCount === 0) return 0;

    return matchSum / required.length;
  }

  /** Find the single best agent for a task, or null if none can match. */
  findBestAgent(
    task: TaskDefinition,
  ): { agentSlug: string; score: number } | null {
    const ranked = this.findCapableAgents(task);
    return ranked.length > 0 ? ranked[0] : null;
  }

  /** Find all capable agents, ranked by match score descending. */
  findCapableAgents(
    task: TaskDefinition,
  ): Array<{ agentSlug: string; score: number }> {
    const results: Array<{ agentSlug: string; score: number }> = [];

    for (const [slug] of this.agents) {
      const score = this.scoreMatch(slug, task);
      if (score > 0) {
        results.push({ agentSlug: slug, score });
      }
    }

    results.sort((a, b) => b.score - a.score);
    return results;
  }
}
